### PR TITLE
Nc 1942 Add account whitelisting and refactor into permissioning package

### PR DESCRIPTION
**Pull request by Shinabyss**
## PR description
Create account whitelisting serving as the groundwork for role based permissioning system for transactions

Include:
- "--accounts-whitelisting" config on startup
- logic to reject transaction on validation when using account whitelist is active
- Unit testing
- refactor account whitelisting from p2p to permissioning package

### CODE REVIEW

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/core/TransactionPoolTest.java [hunk 1]
```diff
@@ -355,6 +359,35 @@ public void shouldNotNotifyBatchListenerIfNoTransactionsAreAdded() {
...
+    when(accountWhitelistController.contains(transaction1.getSender().toString()))
+        .thenReturn(false);
+
+    assertThat(transactionPool.addLocalTransaction(transaction1))
+        .isEqualTo(ValidationResult.invalid(TX_SENDER_NOT_AUTHORIZED));
+
+    assertTransactionNotPending(transaction1);
+    verifyZeroInteractions(batchAddedListener);
+  }
+
```
Comment (lucassaldanha) [inline ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/core/TransactionPoolTest.java]: We should probably have a unit test to verify the behaviour when the `accountWhitelistController` is not present. Something like `shouldAllowTransactionWhenAccountWhitelistControllerIsNotPresent()`.

File: pantheon/src/main/java/tech/pegasys/pantheon/RunnerBuilder.java [hunk 2]
```diff
@@ -221,6 +222,11 @@ public Runner build() {
     final TransactionPool transactionPool = pantheonController.getTransactionPool();
     final MiningCoordinator miningCoordinator = pantheonController.getMiningCoordinator();
 
+    AccountWhitelistController accountWhitelistController =
+        new AccountWhitelistController(permissioningConfiguration);
+
+    transactionPool.setAccountWhitelist(accountWhitelistController);
```
Comment (lucassaldanha) [inline pantheon/src/main/java/tech/pegasys/pantheon/RunnerBuilder.java]: We don't need to call the setter if the account whitelist property hasn't been set. Why don't we check it here?

```
if (permissioningConfiguration.isAccountWhitelistSet()) {
  transactionPool.setAccountWhitelist(accountWhitelistController);
}
```

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/core/TransactionPoolTest.java (hunk 1 above)
Comment (Shinabyss) [reply ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/core/TransactionPoolTest.java]: actioned

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/core/TransactionPool.java [hunk 3]
```diff
@@ -131,6 +134,16 @@ public PendingTransactions getPendingTransactions() {
       return basicValidationResult;
     }
 
+    if (accountWhitelistController.isPresent()
```
[the remark this answers is not in the document]
Comment (Shinabyss) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/core/TransactionPool.java]: actioned

File: pantheon/src/main/java/tech/pegasys/pantheon/RunnerBuilder.java (hunk 2 above)
Comment (Shinabyss) [reply pantheon/src/main/java/tech/pegasys/pantheon/RunnerBuilder.java]: actioned

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/core/TransactionPool.java [hunk 4]
```diff
@@ -131,6 +134,13 @@ public PendingTransactions getPendingTransactions() {
       return basicValidationResult;
     }
 
+    String sender = transaction.getSender().toString();
+    if (!checkIfAccountIsWhitelisted(sender)) {
```
Comment (lucassaldanha) [inline ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/core/TransactionPool.java]: Now that I read this, I think the if statement will be better as something like:

```
if (!accountIsWhitelisted(sender)) {
```

What do you think?

Comment (lucassaldanha) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/core/TransactionPool.java]: Or maybe even `accountIsNotWhitelisted(sender))` so we don't need the negative... idk

Comment (Shinabyss) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/core/TransactionPool.java]: actioned

File: pantheon/src/main/java/tech/pegasys/pantheon/cli/PantheonCommand.java [hunk 5]
```diff
@@ -404,13 +404,24 @@ private void setRefreshDelay(final Long refreshDelay) {
...
+        "Comma separated enode URLs for permissioned networks. You may specify an empty list.",
     split = ",",
     arity = "0..*",
     converter = EnodeToURIPropertyConverter.class
   )
   private final Collection<URI> nodesWhitelist = null;
 
+  @Option(
+    names = {"--accounts-whitelist"},
+    paramLabel = "<hexString of account public key>",
```
[the remark this answers is not in the document]
Comment (Shinabyss) [reply pantheon/src/main/java/tech/pegasys/pantheon/cli/PantheonCommand.java]: actioned

File: pantheon/src/main/java/tech/pegasys/pantheon/cli/PantheonCommand.java [hunk 6]
```diff
@@ -404,13 +404,24 @@ private void setRefreshDelay(final Long refreshDelay) {
...
+    description =
+        "Comma separated hexString of account public key "
+            + "for permissioned/role-based transaction. You may specify an empty list.",
+    split = ",",
```
[the remark this answers is not in the document]
Comment (Shinabyss) [reply pantheon/src/main/java/tech/pegasys/pantheon/cli/PantheonCommand.java]: actioned

File: pantheon/src/main/java/tech/pegasys/pantheon/cli/PantheonCommand.java [hunk 7]
```diff
@@ -404,13 +404,24 @@ private void setRefreshDelay(final Long refreshDelay) {
     names = {"--nodes-whitelist"},
     paramLabel = "<enode://id@host:port>",
     description =
-        "Comma separated enode URLs for permissioned networks. Not intended to be used with mainnet or public testnets.",
+        "Comma separated enode URLs for permissioned networks. You may specify an empty list.",
```
[the remark this answers is not in the document]
Comment (Shinabyss) [reply pantheon/src/main/java/tech/pegasys/pantheon/cli/PantheonCommand.java]: actioned

**Pull request merged**